versioning/spec/versioning.rb

```diff
@@ -0,0 +1,128 @@
+require_relative '../versioning'
+require 'tmpdir'
+
+describe Versioning do
+  def create_commit(file_name)
+    File.write(file_name, 'Dummy content')
+    `git add #{file_name}`
+    `git commit --no-gpg-sign --message Dummy #{file_name}`
+  end
+
+  def create_git_dir_with_tag(tag)
+    `git init`
+    create_commit(tag)
+    `git tag #{tag}`
+  end
+
+  def create_uncomitted_changes(file)
+    File.write(file, 'Dummy content')
+    `git add #{file}`
+  end
+
+  around(:each) do |example|
+    Dir.mktmpdir do |dir|
+      Dir.chdir(dir) do
+        example.run
+      end
+    end
+  end
+
+  describe '.verify_git!' do
+    # Assuming that git is always there when running rspec
+    context 'when git does not exist' do
+      it 'raises an error' do
+        allow(File).to receive(:executable?).and_return(false)
+        expect {
+          Versioning.verify_git!
+        }.to raise_error(StandardError, /The command `git` does not exist!/)
+      end
+    end
+
+    context 'when current is not a git dir' do
+      it 'raises an error' do
+        expect {
+          Versioning.verify_git!
+        }.to raise_error(StandardError, /The current directory `#{Dir.getwd}` is not a git tree!/)
+      end
+    end
+
+    context 'when current path is in a git dir' do
+      it 'does not raise' do
+        `git init`
+        expect {
+          Versioning.verify_git!
+        }.to_not raise_error
+      end
+    end
+  end
+
+  describe '.verify_semver_tags!' do
+    context 'when semver tag exists' do
+      it 'does not raise an error' do
+        create_git_dir_with_tag('v0.0.1')
+        expect {
+          Versioning.verify_semver_tags!
+        }.not_to raise_error
+      end
+    end
+
+    context 'when no semver tag exists' do
+      it 'raises an error' do
+        create_git_dir_with_tag('some_tag')
+        expect {
+          Versioning.verify_semver_tags!
+        }.to raise_error(StandardError, /A git tag with an semantic version is required!/)
+      end
+    end
+
+    context 'when no git tag exists' do
+      it 'raises with the same error message' do
+        `git init`
+        expect {
+          Versioning.verify_semver_tags!
+        }.to raise_error(StandardError, /A git tag with an semantic version is required!/)
+      end
+    end
+  end
+
+  describe '.current_version' do
```

Comment:
I feel like many of the tests here just verify that `git describe` works the way it is documented to work, so not sure of they are strictly necessary, but they don't hurt.

One scenario you may want to add it having a semver tag on an earlier commit, but a non-semver tag afterwards. Since `git describe` doesn't check for semver, it will pick the wrong tag, and this function should fail and not return the non-semver tag.

Reply:
I know but they were created before I was aware that `git describe` already covers the whole specification. And I think if we might change our specification later we can easily adapt the implementation and tests in Ruby.